src/Purebred/UI/Actions.hs

```diff
@@ -216,114 +214,138 @@ first. Keep the 'Action' specific to the view and widget
 
 instance HasList 'ListOfThreads where
   type T 'ListOfThreads = Items
   type E 'ListOfThreads = Toggleable NotmuchThread
-  list _ = asThreadsView . miListOfThreads
+  list = asThreadsView . miListOfThreads
 
 instance HasList 'ListOfMails where
   type T 'ListOfMails = Vector.Vector
   type E 'ListOfMails = Toggleable NotmuchMail
-  list _ = asThreadsView . miListOfMails
+  list = asThreadsView . miListOfMails
 
 instance HasList 'ScrollingMailView where
   type T 'ScrollingMailView = Vector.Vector
   type E 'ScrollingMailView = Toggleable NotmuchMail
-  list _ = asThreadsView . miListOfMails
+  list = asThreadsView . miListOfMails
 
 instance HasList 'ComposeListOfAttachments where
   type T 'ComposeListOfAttachments = Vector.Vector
   type E 'ComposeListOfAttachments = MIMEMessage
-  list _ = asCompose . cAttachments
+  list = asCompose . cAttachments
 
 instance HasList 'MailListOfAttachments where
   type T 'MailListOfAttachments = Vector.Vector
   type E 'MailListOfAttachments = WireEntity
-  list _ = asMailView . mvAttachments
+  list = asMailView . mvAttachments
 
 -- | contexts which have selectable items in a list
 class (HasList (n :: Name), Traversable (T n)) =>
       HasToggleableList n
   where
-  untoggleE :: Proxy n -> E n -> E n
-  toggleE :: Proxy n -> E n -> E n
-  isToggledE :: Proxy n -> E n -> Bool
-  untoggleAll :: (MonadState AppState m) => Proxy n -> m ()
-  untoggleAll proxy = modifying (list proxy . traversed) (untoggleE proxy)
-  toggle :: Proxy n -> Int -> StateT AppState (T.EventM Name) ()
-  -- | Traversal of selected items.  NOT a valid Traversal unless
-  -- selected state is preserved
-  toggledItemsL :: Proxy n -> Traversal' AppState (E n)
-  toggledItemsL proxy = list proxy . traversed . filtered (isToggledE proxy)
+
+  type Inner n
+  toggleState :: Lens' (E n) Bool
+  inner :: Lens' (E n) (Inner n)
+
+  toggle :: StateT AppState (T.EventM Name) ()
+
+untoggleAll :: forall n m. (HasToggleableList n, MonadState AppState m) => m ()
+untoggleAll = assign (list @n . traversed . toggleState @n) False
+
+-- | Traversal of selected items.
+toggledItemsL :: forall n. (HasToggleableList n) => Traversal' AppState (Inner n)
+toggledItemsL = list @n . traversed . filtered (view (toggleState @n)) . inner @n
+
+type family Snd a
+type instance Snd (a, b) = b
 
 instance
   ( HasList n
   , Traversable (T n)
   , E n ~ (Bool, a)
-  , Index (T n (Bool, a)) ~ Int
-  , IxValue (T n (Bool, a)) ~ (Bool, a)
-  , Ixed (T n (Bool, a))
+  , L.Splittable (T n)
+  , Semigroup (T n (Bool, a))
   ) => HasToggleableList n where
-  untoggleE _ = set _1 False
-  toggleE _ = over _1 not
-  isToggledE _ = fst
-  toggle proxy i = modifying (list proxy . L.listElementsL . ix i) (toggleE proxy)
+  type Inner n = Snd (E n)
+  toggleState = _1
+  inner = _2
+  toggle = modifying (list @n . listSelectedElementL . toggleState @n) not
+
+-- | Traversal of selected element (if any)
+--
+-- We should try to upstream this to Brick
```

Comment:
Actually it was already merged: https://github.com/jtdaugherty/brick/pull/382

I mentioned in the commit message but forgot to remove/update this comment.  I'll fix it if I have to rebase or revise this PR.